tools/get_training_info/get_training_info.py

```diff
@@ -0,0 +1,456 @@
+    b_str = ",".join(map(str, b))
+    shape_str = ""
+
+    if w:
+        shape_str += "w=[" if prefix else "["
+        shape_str += w_str + "]"
+    if b:
+        shape_str += ",b=[" + b_str + "]"
+
+    return shape_str.ljust(pad)
+
+
+def get_dist_group_info(args):
+    tp_groups, pp_groups, dp_groups = [], [], []
+
+    num_tensor_model_parallel_groups: int = args.world_size // args.tensor_model_parallel_size
+    num_pipeline_model_parallel_groups: int = args.world_size // args.pipeline_model_parallel_size
+
+    def get_groups(ranks):
+        nodes = [i // args.nproc_per_node for i in ranks]
+        gpus = [i % args.nproc_per_node for i in ranks]
+        groups = [f"n{n_idx}_g{g_idx}" for n_idx, g_idx in zip(nodes, gpus)]
+        return groups
+
+    for i in range(num_tensor_model_parallel_groups):
+        ranks = range(i * args.tensor_model_parallel_size,
+                      (i + 1) * args.tensor_model_parallel_size)
+        tp_groups.append(get_groups(ranks))
+        
+    for i in range(num_pipeline_model_parallel_groups):
+        ranks = range(i, args.world_size, num_pipeline_model_parallel_groups)
+        pp_groups.append(get_groups(ranks))
+
+    for i in range(args.pipeline_model_parallel_size):
+        start_rank = i * num_pipeline_model_parallel_groups
+        end_rank = (i + 1) * num_pipeline_model_parallel_groups
+        for j in range(args.tensor_model_parallel_size):
+            ranks = range(start_rank + j, end_rank,
+                          args.tensor_model_parallel_size)
+            dp_groups.append(get_groups(ranks))
+
+    return tp_groups, pp_groups, dp_groups
+
+
+def get_padded_vocab_size(args):
+    vocab_size = len(json.load(open(args.vocab_file)))
+    multiple = args.make_vocab_size_divisible_by * \
+        args.tensor_model_parallel_size
+    while (vocab_size % multiple) != 0:
+        vocab_size += 1
+    return vocab_size
+
+
+def get_ffn_hidden_size(args):
+    return int((4 * args.hidden_size * 2 / 3) / 64) * 64
+
+
+def get_megatron_info(disable_profile=False):
+    args = get_args()
+    h = args.hidden_size
+    t = args.tensor_model_parallel_size
+    n = args.num_layers
+    p = args.pipeline_model_parallel_size
+    d = args.world_size // p // t
+    v = get_padded_vocab_size(args)
+    s = args.seq_length
+    b = args.micro_batch_size
+    a = args.num_attention_heads
+    ga = args.global_batch_size // args.micro_batch_size // d
+    f = get_ffn_hidden_size(args)
+    per_stage_layer_num = n // p
+
+    # parameters
+    total_parameters = v * h + s * h + (7 * h + 4 * h * h + 3 * f * h +
+                                        2 * f) * n
+    total_parameters_formatted = f'{total_parameters:,}'
+
+    total_parameters_per_gpu = v * h / t + s * h + (
+        7 * h + 4 * h * h / t + 3 * f * h / t + 2 * f) * per_stage_layer_num
+    total_parameters_per_gpu_formatted = f'{int(total_parameters_per_gpu):,}'
+
+    activations = n * (10 * s * h * b +
```

Comment:
Can you explain the formula of computating activation memory so that we can understand the intuition behind this formula?

Reply:
The explanation for this formula can be found in the "Calculation Method Explanation" section